### PR TITLE
Optimise FallbackIfEmpty

### LINKED ISSUE #367: `FallbackIfEmpty` Optimizations
`FallbackIfEmpty` can take advantage of `TryGetCollectionCount`

### PR DESCRIPTION
**Pull request by julienasp**
This PR improves the `FallbackIfEmpty` performance in the case of collections.

Fixes #367

### CODE REVIEW

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs [hunk 1]
```diff
@@ -10,6 +10,14 @@ public void FallbackIfEmptyWithEmptySequence()
 		source.FallbackIfEmpty(12, 23).AssertSequenceEqual(12, 23);
 	}
 
+	[Fact]
+	public void FallbackIfEmptyWithEmptyListSequence()
+	{
+		using var source = Enumerable.Empty<int>().ToList().AsTestingSequence(maxEnumerations: 2);
```
Comment (julienasp) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:16]: not working, because TestingSequence is not recognised by the TryGetCollectionCount, so my test does not go through the new code

Comment (julienasp) [reply Tests/SuperLinq.Test/FallbackIfEmptyTest.cs]: i removed the testing sequence, otherwise the test was not taking the right path.

Might be a good idea to add a CollectionTestingSequence that implements ICollection for all those cases where we use TryGetCollectionCount

**Review by viceroypenguin — COMMENTED** [review]
Check out [this test](https://github.com/viceroypenguin/SuperLinq/blob/2a67bd8c34c8f1b9983c0376270fd638800fa386/Tests/SuperLinq.Test/AssertCountTest.cs#L85) for an example of why we cannot evaluate collection count when the operator is called. 

In further consideration, I wonder if I was forgetting that when I write the issue. In theory, this could be abstracted to an `If()` that selects the right `IEnumerable`, but in practice,  don't know that it's an actual optimization. 

I'm going to leave this open for a few days while I ponder whether to ask you to improve this or to ask me to stop coming up with bad ideas... :)

**Comment by viceroypenguin** [comment]
@julienasp I think I would like to have `FallbackIfEmpty` implemented as a `CollectionIterator`. You can review `DoIterator` for a minimum implementation and `Insert` method for example of how it can be checked for two `IEnumerable<>`.

**Comment by julienasp** [comment]
> @julienasp I think I would like to have `FallbackIfEmpty` implemented as a `CollectionIterator`. You can review `DoIterator` for a minimum implementation and `Insert` method for example of how it can be checked for two `IEnumerable<>`. I think the `GetEnumerable()` and `CopyTo()` methods can be implemented.

perfect i'll work on that this weekend

File: Source/SuperLinq/FallbackIfEmpty.cs [hunk 2]
```diff
@@ -40,16 +40,21 @@ public static IEnumerable<T> FallbackIfEmpty<T>(this IEnumerable<T> source, IEnu
 		Guard.IsNotNull(source);
 		Guard.IsNotNull(fallback);
 
-		return Core(source, fallback);
+		return source is ICollection<T>
```
Comment (viceroypenguin) [inline Source/SuperLinq/FallbackIfEmpty.cs:43]:
```suggestion
		return source.TryGetCollectionCount() is not null
```

File: Source/SuperLinq/FallbackIfEmpty.cs [hunk 3]
```diff
@@ -58,4 +63,28 @@ static IEnumerable<T> Core(IEnumerable<T> source, IEnumerable<T> fallback)
...
+		private readonly IEnumerable<T> _fallback;
+
+		public FallbackIfEmptyCollectionIterator(IEnumerable<T> source, IEnumerable<T> fallback)
+		{
+			_source = source;
+			_fallback = fallback;
+		}
+
+		public override int Count =>
+			_source.TryGetCollectionCount() is { } and 0
```
Comment (viceroypenguin) [inline Source/SuperLinq/FallbackIfEmpty.cs:79]:
```suggestion
			_source.GetCollectionCount() == 0
```

File: Source/SuperLinq/FallbackIfEmpty.cs [hunk 4]
```diff
@@ -58,4 +63,28 @@ static IEnumerable<T> Core(IEnumerable<T> source, IEnumerable<T> fallback)
...
+				? _fallback.Count()
+				: _source.GetCollectionCount();
+
+		protected override IEnumerable<T> GetEnumerable()
+		{
+			return _source.TryGetCollectionCount() is { } and 0
```
Comment (viceroypenguin) [inline Source/SuperLinq/FallbackIfEmpty.cs:85]:
```suggestion
			return _source.TryGetCollectionCount() == 0
```

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:1]: Please add `CollectionBehavior` tests (one for empty, one for non-empty); see `AssertCountCollectionBehavior` for example.

Also, please add `UsesCollectionCountAtIterationTime` test; see `AssertCountUsesCollectionCountAtIterationTime` for example. Instead of 3->4, it would be 0->1.

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs [hunk 5]
```diff
@@ -10,6 +10,22 @@ public void FallbackIfEmptyWithEmptySequence()
 		source.FallbackIfEmpty(12, 23).AssertSequenceEqual(12, 23);
 	}
 
+	[Fact]
+	public void FallbackIfEmptyWithEmptyListSequence()
```
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:14]: Test is unnecessary; version with `AsTestingCollection()` is appropriate to cover.

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs [hunk 6]
```diff
@@ -10,6 +10,22 @@ public void FallbackIfEmptyWithEmptySequence()
...
+	{
+		var source = Enumerable.Empty<int>().ToList();
+		source.FallbackIfEmpty(12).AssertSequenceEqual(12);
+		source.FallbackIfEmpty(12, 23).AssertSequenceEqual(12, 23);
+	}
+
+	[Fact]
+	public void FallbackIfEmptyWithCollectionListSequence()
```
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:22]: Add a matching test with a non-empty collection, which passes in a `BreakingSequence`.

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:1]: This is not your fault, but while we're fixing things... Please update `FallbackIfEmptyWithNotEmptySequence` to use `BreakingSequence` instead of `12, 23`. Also please update all tests to only do a single example instead of two. The dual-check on each is unnecessary.

**Review by viceroypenguin — COMMENTED** [review]
Thanks! Looking good so far.

File: Source/SuperLinq/FallbackIfEmpty.cs [hunk 7]
```diff
@@ -40,16 +40,21 @@ public static IEnumerable<T> FallbackIfEmpty<T>(this IEnumerable<T> source, IEnu
 		Guard.IsNotNull(source);
 		Guard.IsNotNull(fallback);
 
-		return Core(source, fallback);
+		return source.TryGetCollectionCount() is not null
```
Comment (viceroypenguin) [inline Source/SuperLinq/FallbackIfEmpty.cs:43]: `Count` as a property should not require enumeration to run. As such, we need to enforce that both `source` and `fallback` are valid before passing to the collection iterator: 
```suggestion
		return source.TryGetCollectionCount() is not null 
			&& fallback.TryGetCollectionCount() is not null
```

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs [hunk 8]
```diff
@@ -7,14 +7,58 @@ public void FallbackIfEmptyWithEmptySequence()
 	{
 		using var source = Enumerable.Empty<int>().AsTestingSequence(maxEnumerations: 2);
 		source.FallbackIfEmpty(12).AssertSequenceEqual(12);
-		source.FallbackIfEmpty(12, 23).AssertSequenceEqual(12, 23);
+	}
+
+	[Fact]
+	public void FallbackIfEmptyWithCollectionSequence()
+	{
+		using var source = Enumerable.Empty<int>().AsTestingCollection(maxEnumerations: 2);
```
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:15]: `maxEnumerations` does not need to be set to `2`, now that we are not executing `FallbackIfEmpty()` twice.

File: Tests/SuperLinq.Test/FallbackIfEmptyTest.cs [hunk 9]
```diff
@@ -7,14 +7,58 @@ public void FallbackIfEmptyWithEmptySequence()
...
+		using var source = Seq(1).AsTestingCollection(maxEnumerations: 2);
+
+		var result = source.FallbackIfEmpty(new BreakingSequence<int>());
+		result.AssertCollectionErrorChecking(1);
+	}
+
+	[Fact]
+	public void FallbackIfEmptyUsesCollectionCountAtIterationTime()
+	{
+		using var source = new BreakingCountCollection<int>();
```
Comment (viceroypenguin) [inline Tests/SuperLinq.Test/FallbackIfEmptyTest.cs:54]: We don't need to use `BreakingCountCollection` here. Simply follow the example in other `CollectionCountAtIterationTime` methods. Create a stack and set up the result with a sequence. Test that the result sequence is the fallback sequence. Then add a value to the stack, and evaluate result again that it uses the stack sequence instead of the fallback sequence.

The corner case we are testing here is that the behavior of `FallbackIfEmpty` is not determined by the value of `Count` when `FallbackIfEmpty` is called, but when it is enumerated. See morelinq/morelinq#945 for explanation of error sequence. I see that we are evaluating correctly here, but we should validate the future changes don't break this behavior.

**Review by viceroypenguin — COMMENTED** [review]
I think this is the last round of changes. 🤞🏼

**Comment by viceroypenguin** [comment]
@julienasp Thanks!
